diff -urN network-2.5.0.0/cbits/swap.c network-2.5.0.0-new/cbits/swap.c
--- network-2.5.0.0/cbits/swap.c	1970-01-01 00:00:00.000000000 +0000
+++ network-2.5.0.0-new/cbits/swap.c	2014-11-01 19:23:03.650641408 +0000
@@ -0,0 +1,16 @@
+#include <arpa/inet.h>
+
+static uint32_t my_htonl(uint32_t x) { return htonl(x); }
+static uint16_t my_htons(uint16_t x) { return htons(x); }
+static uint32_t my_ntohl(uint32_t x) { return ntohl(x); }
+static uint16_t my_ntohs(uint16_t x) { return ntohs(x); }
+
+#undef htonl
+#undef htons
+#undef ntohl
+#undef ntohs
+
+uint32_t htonl(uint32_t x) { return my_htonl(x); }
+uint16_t htons(uint16_t x) { return my_htons(x); }
+uint32_t ntohl(uint32_t x) { return my_ntohl(x); }
+uint16_t ntohs(uint16_t x) { return my_ntohs(x); }
diff -urN network-2.5.0.0/network.cabal network-2.5.0.0-new/network.cabal
--- network-2.5.0.0/network.cabal	2014-11-01 19:25:00.895746048 +0000
+++ network-2.5.0.0-new/network.cabal	2014-11-01 19:27:51.634388480 +0000
@@ -47,6 +47,10 @@
   if !os(windows)
     build-depends:
       unix >= 2
+  if os(haiku)
+    extra-libraries:
+      network
+    c-sources: cbits/swap.c
 
   extensions:
     CPP, DeriveDataTypeable, ForeignFunctionInterface, TypeSynonymInstances
